janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/backpressure/SemaphoreQueryBackPressure.java

```diff
@@ -0,0 +1,65 @@
+// Copyright 2023 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.util.backpressure;
+
+import org.janusgraph.core.JanusGraphException;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+
+import static org.janusgraph.util.system.ExecuteUtil.gracefulExecutorServiceShutdown;
+
+public class SemaphoreQueryBackPressure implements QueryBackPressure{
+
+    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+    private final Runnable releaseNonBlocking;
+    private final Semaphore semaphore;
+
+    public SemaphoreQueryBackPressure(final int backPressureLimit) {
+        this.semaphore = new Semaphore(backPressureLimit, true);
+        this.releaseNonBlocking = () -> {
+            // ensure we never add more permits than `backPressureLimit`
+            // (even if `releaseAfterQuery()` is called more times than `acquireBeforeQuery()`);
+            if(semaphore.availablePermits()<backPressureLimit){
```

Comment:
Actually this shouldn't ever be evaluated to `false` unless I made a bug anywhere. I was checking multiple times for cases when this would be evaluated to `false` but couldn't find any potentially possible case yet.
I thought also that this `SemaphoreQueryBackPressure` may potentially be reused for other storage backend implementations (like `HBase` for example) and we could later introduce some kind of a bug when we accidentally release more times that needed which would result to a bug which is pretty hard to debug / find.
This might be overprotective methods I guess and if we decide to remove this check then we can also remove the executorService here as it won't be necessary anymore (related to your next [comment](https://github.com/JanusGraph/janusgraph/pull/3760#discussion_r1190666084)).
Currently for CQL `getSlice` function implementation I changed the logic to retrieve the next page while we are processing the current page (for performance improvements). I thought that potentially there could be some bug where we fail to process current page and then the next page retrieval is failed as well (for whatever reason) and we accidentally release it 2 times instead of releasing only once. That said, I checked the logic over and over again and couldn't find a case where we would potentially trigger release more than once for a single `acquire`. 
I tried to ensure that we always trigger single `release` for single `acquire` (no more than no less). 
That said, I do understand that I could potentially make a bug which I don't see right now. In case I made a bug (or in case someone uses it later and makes a bug) then releasing more (or less) then necessary could be quite a hard bug to find because it will mean that the BackPressure limit is now changed and from now on we will send more or less queries in parallel then necessary.
In case we send more queries in parallel then the configured back pressure limit then it could result in underlying driver overload which will cause with exceptions during queries evaluation.
In case the limit is now changed to be less, then we will send less queries in parallel which may result in degraded performance.
Thus, it's very important to ensure that for a single `acquire` we always trigger a single `release` later (and only single `release`).
I'm thinking that we could / should potentially log a warning / error if `semaphore.availablePermits()<backPressureLimit` is evaluated to false to bring user attention to this and ask them to reach JanusGraph community for the bug investigation.
Another solution would be to remove this check, remove `executorService` and just add more Javadoc, so that developers who want to reuse `SemaphoreQueryBackPressure` later would understand that executing `release` more or less than once for any `acquire` will result in bugs which are hard to be detected and found.

Reply:
In short - this check is here for extra safety measures only. If you think we shouldn't have this check here, I'm good with removing it which will also remove the `executorService` and we will use direct `release` call for the underlying Semaphore.

Reply:
What you said makes sense - but I feel like it's an overkill. I never read `Semaphore`'s documentation carefully. I would like to see if some one else could chime in.

Reply:
> I'm thinking that we could / should potentially log a warning / error if semaphore.availablePermits()<backPressureLimit is evaluated to false to bring user attention to this and ask them to reach JanusGraph community for the bug investigation

Is it possible to make this testing-only, not enabled by default, and add some high concurrent tests? We may not have the infrastructure for that yet, so I am just brainstorming.

Reply:
I think it makes sense to allow users to use their own back pressure implementation if they don't like Semaphore based implementation. Moreover, knowing that this back pressure mechanisms may be reused for other backend implementations, I think it make sense to make it more expansible.
 I added 3 back pressure implementations:
- `PassAllQueryBackPressure` - back pressure is not used.
- `SemaphoreQueryBackPressure` - back pressure implementation using `Semaphore` (without unnecessary checks or executor service).
- `SemaphoreProtectedReleaseQueryBackPressure` - back pressure implementation using `Semaphore` where `release` is asynchronous / non-blocking and processed by a singly threaded `ExecutorService`.

I added Javadoc for all the implementations to know when it's best to use them. By default I selected `SemaphoreQueryBackPressure` to be used because it has smaller overhead. 

I tried to benchmark releasing logic with gradually increasing resources usage and changing back pressure limits for all 3 implementations. Moreover for `SemaphoreProtectedReleaseQueryBackPressure` I tested both awaiting for release to actually happen and not waiting for release to happen (just submitting the release job).

The final results are here:
```
Benchmark                             (backPressure)  (threads)                                                     (type)  Mode  Cnt    Score    Error  Units
BackPressureBenchmark.releaseBlocked           50000       2000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5  110.045 ± 10.087  ms/op
BackPressureBenchmark.releaseBlocked           50000       2000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5  112.751 ±  9.358  ms/op
BackPressureBenchmark.releaseBlocked           50000       2000                                      semaphoreBackPressure  avgt    5  113.015 ± 11.294  ms/op
BackPressureBenchmark.releaseBlocked           50000       2000                                        passAllBackPressure  avgt    5  109.984 ± 13.694  ms/op

BackPressureBenchmark.releaseBlocked           50000       1000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5   54.572 ±  7.032  ms/op
BackPressureBenchmark.releaseBlocked           50000       1000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5   53.349 ±  5.845  ms/op
BackPressureBenchmark.releaseBlocked           50000       1000                                      semaphoreBackPressure  avgt    5   52.626 ±  5.060  ms/op
BackPressureBenchmark.releaseBlocked           50000       1000                                        passAllBackPressure  avgt    5   51.725 ± 10.173  ms/op

BackPressureBenchmark.releaseBlocked           50000        100     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    4.780 ±  0.188  ms/op
BackPressureBenchmark.releaseBlocked           50000        100  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    4.789 ±  0.389  ms/op
BackPressureBenchmark.releaseBlocked           50000        100                                      semaphoreBackPressure  avgt    5    4.689 ±  0.338  ms/op
BackPressureBenchmark.releaseBlocked           50000        100                                        passAllBackPressure  avgt    5    4.524 ±  0.422  ms/op

BackPressureBenchmark.releaseBlocked           50000         10     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.316 ±  0.015  ms/op
BackPressureBenchmark.releaseBlocked           50000         10  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.303 ±  0.020  ms/op
BackPressureBenchmark.releaseBlocked           50000         10                                      semaphoreBackPressure  avgt    5    0.275 ±  0.030  ms/op
BackPressureBenchmark.releaseBlocked           50000         10                                        passAllBackPressure  avgt    5    0.266 ±  0.016  ms/op

BackPressureBenchmark.releaseBlocked           50000          4     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.186 ±  0.030  ms/op
BackPressureBenchmark.releaseBlocked           50000          4  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.165 ±  0.009  ms/op
BackPressureBenchmark.releaseBlocked           50000          4                                      semaphoreBackPressure  avgt    5    0.146 ±  0.016  ms/op
BackPressureBenchmark.releaseBlocked           50000          4                                        passAllBackPressure  avgt    5    0.142 ±  0.014  ms/op

BackPressureBenchmark.releaseBlocked           50000          2     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.145 ±  0.031  ms/op
BackPressureBenchmark.releaseBlocked           50000          2  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.119 ±  0.019  ms/op
BackPressureBenchmark.releaseBlocked           50000          2                                      semaphoreBackPressure  avgt    5    0.107 ±  0.015  ms/op
BackPressureBenchmark.releaseBlocked           50000          2                                        passAllBackPressure  avgt    5    0.107 ±  0.020  ms/op

BackPressureBenchmark.releaseBlocked           50000          1     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.147 ±  0.017  ms/op
BackPressureBenchmark.releaseBlocked           50000          1  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.115 ±  0.012  ms/op
BackPressureBenchmark.releaseBlocked           50000          1                                      semaphoreBackPressure  avgt    5    0.072 ±  0.014  ms/op
BackPressureBenchmark.releaseBlocked           50000          1                                        passAllBackPressure  avgt    5    0.071 ±  0.010  ms/op

BackPressureBenchmark.releaseBlocked           10000       2000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5  115.985 ± 15.574  ms/op
BackPressureBenchmark.releaseBlocked           10000       2000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5  120.029 ± 19.969  ms/op
BackPressureBenchmark.releaseBlocked           10000       2000                                      semaphoreBackPressure  avgt    5  116.981 ± 12.926  ms/op
BackPressureBenchmark.releaseBlocked           10000       2000                                        passAllBackPressure  avgt    5  111.556 ± 16.965  ms/op

BackPressureBenchmark.releaseBlocked           10000       1000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5   54.079 ±  7.095  ms/op
BackPressureBenchmark.releaseBlocked           10000       1000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5   52.309 ±  3.651  ms/op
BackPressureBenchmark.releaseBlocked           10000       1000                                      semaphoreBackPressure  avgt    5   54.030 ±  2.800  ms/op
BackPressureBenchmark.releaseBlocked           10000       1000                                        passAllBackPressure  avgt    5   51.631 ±  4.806  ms/op

BackPressureBenchmark.releaseBlocked           10000        100     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    4.720 ±  0.114  ms/op
BackPressureBenchmark.releaseBlocked           10000        100  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    4.764 ±  0.211  ms/op
BackPressureBenchmark.releaseBlocked           10000        100                                      semaphoreBackPressure  avgt    5    4.729 ±  0.575  ms/op
BackPressureBenchmark.releaseBlocked           10000        100                                        passAllBackPressure  avgt    5    4.628 ±  0.326  ms/op

BackPressureBenchmark.releaseBlocked           10000         10     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.315 ±  0.024  ms/op
BackPressureBenchmark.releaseBlocked           10000         10  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.305 ±  0.011  ms/op
BackPressureBenchmark.releaseBlocked           10000         10                                      semaphoreBackPressure  avgt    5    0.268 ±  0.013  ms/op
BackPressureBenchmark.releaseBlocked           10000         10                                        passAllBackPressure  avgt    5    0.272 ±  0.036  ms/op

BackPressureBenchmark.releaseBlocked           10000          4     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.169 ±  0.017  ms/op
BackPressureBenchmark.releaseBlocked           10000          4  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.149 ±  0.023  ms/op
BackPressureBenchmark.releaseBlocked           10000          4                                      semaphoreBackPressure  avgt    5    0.126 ±  0.015  ms/op
BackPressureBenchmark.releaseBlocked           10000          4                                        passAllBackPressure  avgt    5    0.124 ±  0.018  ms/op

BackPressureBenchmark.releaseBlocked           10000          2     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.130 ±  0.023  ms/op
BackPressureBenchmark.releaseBlocked           10000          2  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.094 ±  0.023  ms/op
BackPressureBenchmark.releaseBlocked           10000          2                                      semaphoreBackPressure  avgt    5    0.085 ±  0.011  ms/op
BackPressureBenchmark.releaseBlocked           10000          2                                        passAllBackPressure  avgt    5    0.084 ±  0.016  ms/op

BackPressureBenchmark.releaseBlocked           10000          1     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.129 ±  0.028  ms/op
BackPressureBenchmark.releaseBlocked           10000          1  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.098 ±  0.036  ms/op
BackPressureBenchmark.releaseBlocked           10000          1                                      semaphoreBackPressure  avgt    5    0.055 ±  0.011  ms/op
BackPressureBenchmark.releaseBlocked           10000          1                                        passAllBackPressure  avgt    5    0.061 ±  0.024  ms/op

BackPressureBenchmark.releaseBlocked            1000       2000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5  119.513 ± 30.644  ms/op
BackPressureBenchmark.releaseBlocked            1000       2000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5  118.678 ± 25.784  ms/op
BackPressureBenchmark.releaseBlocked            1000       2000                                      semaphoreBackPressure  avgt    5  113.144 ± 15.864  ms/op
BackPressureBenchmark.releaseBlocked            1000       2000                                        passAllBackPressure  avgt    5  112.086 ±  5.645  ms/op

BackPressureBenchmark.releaseBlocked            1000       1000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5   51.965 ±  5.159  ms/op
BackPressureBenchmark.releaseBlocked            1000       1000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5   52.599 ±  2.892  ms/op
BackPressureBenchmark.releaseBlocked            1000       1000                                      semaphoreBackPressure  avgt    5   54.557 ±  5.771  ms/op
BackPressureBenchmark.releaseBlocked            1000       1000                                        passAllBackPressure  avgt    5   51.094 ±  5.218  ms/op

BackPressureBenchmark.releaseBlocked            1000        100     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    4.662 ±  0.227  ms/op
BackPressureBenchmark.releaseBlocked            1000        100  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    4.763 ±  0.197  ms/op
BackPressureBenchmark.releaseBlocked            1000        100                                      semaphoreBackPressure  avgt    5    4.645 ±  0.632  ms/op
BackPressureBenchmark.releaseBlocked            1000        100                                        passAllBackPressure  avgt    5    4.653 ±  0.644  ms/op

BackPressureBenchmark.releaseBlocked            1000         10     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.326 ±  0.057  ms/op
BackPressureBenchmark.releaseBlocked            1000         10  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.323 ±  0.048  ms/op
BackPressureBenchmark.releaseBlocked            1000         10                                      semaphoreBackPressure  avgt    5    0.283 ±  0.029  ms/op
BackPressureBenchmark.releaseBlocked            1000         10                                        passAllBackPressure  avgt    5    0.269 ±  0.021  ms/op

BackPressureBenchmark.releaseBlocked            1000          4     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.160 ±  0.024  ms/op
BackPressureBenchmark.releaseBlocked            1000          4  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.137 ±  0.030  ms/op
BackPressureBenchmark.releaseBlocked            1000          4                                      semaphoreBackPressure  avgt    5    0.123 ±  0.019  ms/op
BackPressureBenchmark.releaseBlocked            1000          4                                        passAllBackPressure  avgt    5    0.112 ±  0.018  ms/op

BackPressureBenchmark.releaseBlocked            1000          2     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.127 ±  0.025  ms/op
BackPressureBenchmark.releaseBlocked            1000          2  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.095 ±  0.022  ms/op
BackPressureBenchmark.releaseBlocked            1000          2                                      semaphoreBackPressure  avgt    5    0.083 ±  0.015  ms/op
BackPressureBenchmark.releaseBlocked            1000          2                                        passAllBackPressure  avgt    5    0.081 ±  0.009  ms/op

BackPressureBenchmark.releaseBlocked            1000          1     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.116 ±  0.016  ms/op
BackPressureBenchmark.releaseBlocked            1000          1  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.089 ±  0.019  ms/op
BackPressureBenchmark.releaseBlocked            1000          1                                      semaphoreBackPressure  avgt    5    0.058 ±  0.007  ms/op
BackPressureBenchmark.releaseBlocked            1000          1                                        passAllBackPressure  avgt    5    0.054 ±  0.016  ms/op

BackPressureBenchmark.releaseBlocked             100       2000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5  114.037 ± 12.145  ms/op
BackPressureBenchmark.releaseBlocked             100       2000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5  110.481 ± 10.581  ms/op
BackPressureBenchmark.releaseBlocked             100       2000                                      semaphoreBackPressure  avgt    5  109.492 ± 21.516  ms/op
BackPressureBenchmark.releaseBlocked             100       2000                                        passAllBackPressure  avgt    5  113.730 ± 15.997  ms/op

BackPressureBenchmark.releaseBlocked             100       1000     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5   52.794 ±  4.663  ms/op
BackPressureBenchmark.releaseBlocked             100       1000  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5   53.262 ±  6.986  ms/op
BackPressureBenchmark.releaseBlocked             100       1000                                      semaphoreBackPressure  avgt    5   55.480 ±  7.920  ms/op
BackPressureBenchmark.releaseBlocked             100       1000                                        passAllBackPressure  avgt    5   51.758 ±  5.032  ms/op

BackPressureBenchmark.releaseBlocked             100        100     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    4.653 ±  0.172  ms/op
BackPressureBenchmark.releaseBlocked             100        100  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    4.732 ±  0.112  ms/op
BackPressureBenchmark.releaseBlocked             100        100                                      semaphoreBackPressure  avgt    5    4.576 ±  0.277  ms/op
BackPressureBenchmark.releaseBlocked             100        100                                        passAllBackPressure  avgt    5    4.586 ±  0.576  ms/op

BackPressureBenchmark.releaseBlocked             100         10     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.318 ±  0.012  ms/op
BackPressureBenchmark.releaseBlocked             100         10  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.302 ±  0.034  ms/op
BackPressureBenchmark.releaseBlocked             100         10                                      semaphoreBackPressure  avgt    5    0.270 ±  0.026  ms/op
BackPressureBenchmark.releaseBlocked             100         10                                        passAllBackPressure  avgt    5    0.266 ±  0.027  ms/op

BackPressureBenchmark.releaseBlocked             100          4     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.157 ±  0.030  ms/op
BackPressureBenchmark.releaseBlocked             100          4  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.136 ±  0.032  ms/op
BackPressureBenchmark.releaseBlocked             100          4                                      semaphoreBackPressure  avgt    5    0.118 ±  0.011  ms/op
BackPressureBenchmark.releaseBlocked             100          4                                        passAllBackPressure  avgt    5    0.116 ±  0.019  ms/op

BackPressureBenchmark.releaseBlocked             100          2     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.127 ±  0.019  ms/op
BackPressureBenchmark.releaseBlocked             100          2  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.095 ±  0.013  ms/op
BackPressureBenchmark.releaseBlocked             100          2                                      semaphoreBackPressure  avgt    5    0.082 ±  0.010  ms/op
BackPressureBenchmark.releaseBlocked             100          2                                        passAllBackPressure  avgt    5    0.081 ±  0.006  ms/op

BackPressureBenchmark.releaseBlocked             100          1     semaphoreReleaseProtectedBackPressureWithReleasesAwait  avgt    5    0.121 ±  0.016  ms/op
BackPressureBenchmark.releaseBlocked             100          1  semaphoreReleaseProtectedBackPressureWithoutReleasesAwait  avgt    5    0.089 ±  0.011  ms/op
BackPressureBenchmark.releaseBlocked             100          1                                      semaphoreBackPressure  avgt    5    0.056 ±  0.012  ms/op
BackPressureBenchmark.releaseBlocked             100          1                                        passAllBackPressure  avgt    5    0.058 ±  0.006  ms/op
``` 

The benchmark showed the next:
`PassAllQueryBackPressure` is the fastest implementation but `SemaphoreQueryBackPressure` has almost the same performance in all cases (1% to 5% difference). `SemaphoreProtectedReleaseQueryBackPressure` seems to be almost 50% slower in singe threaded environment, but the more threads / load we have the less difference is noticeable. When there are only 10 threads running the difference is only 15% when compared `SemaphoreProtectedReleaseQueryBackPressure` to `SemaphoreQueryBackPressure`. When there are 100 threads running the difference is not noticeable anymore. 
I'm not sure that the added benchmark is meaningful or correct but the results make sense to me. 
I'm almost sure that back pressure `release` method execution won't be a bottleneck because in comparison it takes for a CQL query to be sent and received the overhead is negligible. 
I propose `SemaphoreQueryBackPressure` to be a default implementation as it has the smallest overhead from two Semaphore based implementations. 
If some developers want to reuse back pressure mechanisms for other storage backend implementations they will be able to read Javadoc of  `SemaphoreQueryBackPressure` which tells about potential wrong usage of this implementation.

@li-boxuan I added another commit to this PR with this implementation / benchmark / tests https://github.com/JanusGraph/janusgraph/pull/3760/commits/0134eb9ca2dc66b5e3429e5b533adc4a4f60ead1 . Let me know what you think about it (I will squash it before merging this PR).

Reply:
> Is it possible to make this check testing-only, and add some high concurrent tests? We may not have the infrastructure for that yet, so I am just brainstorming.

This check is now implemented in `SemaphoreProtectedReleaseQueryBackPressure` only and it's used for tests only.
The default implementation `SemaphoreQueryBackPressure` doesn't have this check anymore.
Benchmark test `BackPressureBenchmark` (excluded by default).

Reply:
Sounds great!